perses/rjmc/topology_proposal.py

```diff
@@ -238,10 +299,394 @@ def propose(self, current_system, current_topology, current_positions, current_m
+
+        return(modeller, missing_atoms)
+
+    def _to_delete(self, modeller, delete_atoms):
+        """
+        remove instances of atoms and corresponding bonds from modeller
+
+        Arguments
+        ---------
+        modeller : simtk.openmm.app.Modeller
+        delete_atoms : list(simtk.openmm.app.topology.Atom) 
+            atoms from existing residue not in new residue
+        Returns
+        -------
+        modeller : simtk.openmm.app.Modeller 
+            extra atoms from old residue have been deleted, missing atoms in new residue not yet added
+        """
+        # delete_atoms : list(simtk.openmm.app.topology.Atom) atoms from existing residue not in new residue
+        # atom : simtk.openmm.app.topology.Atom
+        for atom in delete_atoms:
+            atom.residue._atoms.remove(atom)
+            for bond in modeller.topology._bonds:
+                if atom in bond:
+                    modeller.topology._bonds = filter(lambda a: a != bond, modeller.topology._bonds)
+        # modeller : simtk.openmm.app.Modeller extra atoms from old residue have been deleted, missing atoms in new residue not yet added
+        return modeller
+
+    def _to_delete_bonds(self, modeller, residue_map):
+        """
+        Remove any bonds between atoms in both new and old residue that do not belong in new residue
+        (e.g. breaking the ring in PRO)
+        Arguments
+        ---------
+        modeller : simtk.openmm.app.Modeller
+        residue_map : list(tuples)
+            simtk.openmm.app.topology.Residue (existing residue), str (three letter residue name of proposed residue)
+        Returns
+        -------
+        modeller : simtk.openmm.app.Modeller 
+            extra atoms and bonds from old residue have been deleted, missing atoms in new residue not yet added
+        """
+ 
+        for residue, replace_with in residue_map:
+            # template : simtk.openmm.app.topology._TemplateData
+            template = self._templates[replace_with]
+
+            old_res_bonds = list()
+            # bond : tuple(simtk.openmm.app.topology.Atom, simtk.openmm.app.topology.Atom)
+            for atom1, atom2 in modeller.topology._bonds:
+                if atom1.residue == residue or atom2.residue == residue:
+                    old_res_bonds.append((atom1.name, atom2.name))
+            # make a list of bonds that should exist in new residue
+            # template_bonds : list(tuple(str (atom name), str (atom name))) bonds in template
+            template_bonds = [(template.atoms[bond[0]].name, template.atoms[bond[1]].name) for bond in template.bonds]
+            # add any bonds that exist in template but not in new residue
+            for bond in old_res_bonds:
+                if bond not in template_bonds and (bond[1],bond[0]) not in template_bonds:
+                    modeller.topology._bonds = filter(lambda a: a != bond, modeller.topology._bonds)
```

Comment:
Ughhhhhh python 2 vs 3 problem, good catch -- this also won't solve the test thing, since it was still ```_bonds``` was still a list for me